src/Storages/Distributed/DirectoryMonitor.cpp

```diff
@@ -311,6 +311,35 @@ void StorageDistributedDirectoryMonitor::processFile(const std::string & file_pa
     LOG_TRACE(log, "Started processing `{}`", file_path);
     auto timeouts = ConnectionTimeouts::getTCPTimeoutsWithFailover(storage.global_context.getSettingsRef());
 
+    /// Check that the file is valid
+    /// FIXME: suboptimal
```

Comment:
Another way is to store the size of the compressed file and number of rows in it in the DistributedBlockOutputStream::writeToShard, and then decompression code path can be removed completely (not only this place but also processFilesWithBatches) from the DirectoryMonitory

Reply:
Yes, we need to check it without decompression.
CheckingCompressedReadBuffer will be Ok to implement.
The format was designed to allow sending pre-compressed data.

Reply:
But it will still require calculating hashes, so reading the whole file.

I was thinking about adding file size to the header of the .bin file and verify only it, thus we can avoid reading the whole file.
Since we only need to protect from the truncation on the sender, the receiver will check per-compressed block checksums anyway and fail with an error.

Reply:
Yeah, may be we can store new format as xxx.bin2 and use that double-read approach with unpacking / checksum checking for old format (bin).

Reply:
> But it will still require calculating hashes, so reading the whole file.

Maybe do it while sending?

Reply:
>Maybe do it while sending?

@alexey-milovidov the receiver will check per-block checksum anyway, are you suggesting checking them on the sender? This will complicated the code, and I'm not sure that this worth it.

Reply:
>Yeah, may be we can store new format as xxx.bin2 and use that double-read approach with unpacking / checksum checking for old format (bin).

@filimonov The format already have the header, no need to introduce another extension.

Reply:
> the receiver will check per-block checksum anyway, are you suggesting checking them on the sender?

Yes.

> This will complicated the code, and I'm not sure that this worth it.

Then it's not necessarily.
But we should not read data twice.

If you can check the size it will be enough.

Reply:
> @filimonov The format already have the header, no need to introduce another extension

Backward  compatibility and UX during version upgrade concerns: Will both new a old versions will know how to read files with old/new header? BTW there is also File format to read files from Distributed folder AFAIR.

Reply:
>CheckingCompressedReadBuffer will be Ok to implement.

@alexey-milovidov I went this way, since size of the file cannot be stored in the header (without trickery), although it can be stored in the footer, but I decided that implementing CheckingCompressedReadBuffer will be better (it will verify checksums and avoid seekign back and forth in the file, since after seeking to `file_size-sizeof(footer)` you need to seek back or open two file descriptors).

Reply:
>Backward compatibility and UX during version upgrade concerns: Will both new and old versions know how to read files with old/new header?

@filimonov Yes

Current version changes the on disk format, but not for checksums, but to include rows/bytes to avoid reading the full file to obtain this information (it is required to do proper batching), and if this information does not exists in the header it will be calculated as before (by reading whole file).

>BTW there is also File format to read files from Distributed folder AFAIR.

It uses the same code so it should be fully compatible.